src/map/clif.cpp

```diff
@@ -3208,6 +3208,122 @@ void clif_guild_xy_remove(struct map_session_data *sd)
 	clif_send(buf,packet_len(0x1eb),&sd->bl,GUILD_SAMEMAP_WOS);
 }
 
+void clif_guild_castle_list(struct map_session_data* sd)
+{
+#if PACKETVER_MAIN_NUM >= 20190731 || PACKETVER_RE_NUM >= 20190717 || PACKETVER_ZERO_NUM >= 20190814
+	nullpo_retv(sd);
+
+	struct guild* g = sd->guild;
+	if (g == nullptr)
+		return;
+
+	int castle_count = guild_checkcastles(g);
+
+	if (castle_count > 0) {
+		int len = sizeof(struct PACKET_ZC_GUILD_CASTLE_LIST) + castle_count;
+		struct PACKET_ZC_GUILD_CASTLE_LIST* p = (struct PACKET_ZC_GUILD_CASTLE_LIST*)packet_buffer;
+		p->packetType = HEADER_ZC_GUILD_CASTLE_LIST;
+		p->packetLength = len;
+
+		int i = 0;
+		for (const auto& gc : castle_db) {
+			if (gc.second->guild_id == g->guild_id) {
+				p->castle_list[i] = gc.second->castle_id;
+				++i;
+			}
+		}
+
+		clif_send(p, len, &sd->bl, SELF);
+	}
+#endif
+}
+
+void clif_guild_castleinfo(struct map_session_data* sd, int castle_id, int economy, int defense)
+{
+#if PACKETVER_MAIN_NUM >= 20190731 || PACKETVER_RE_NUM >= 20190717 || PACKETVER_ZERO_NUM >= 20190814
+
+	nullpo_retv(sd);
+
+	struct PACKET_ZC_CASTLE_INFO p;
+	p.packetType = HEADER_ZC_CASTLE_INFO;
+	p.castle_id = castle_id;
+	p.economy = economy;
+	p.defense = defense;
+	clif_send(&p, sizeof(p), &sd->bl, SELF);
+#endif
+}
+
+static void clif_guild_castle_teleport_res(struct map_session_data* sd, enum e_siege_teleport_result result)
+{
+#if PACKETVER_MAIN_NUM >= 20190731 || PACKETVER_RE_NUM >= 20190717 || PACKETVER_ZERO_NUM >= 20190814
+
+	nullpo_retv(sd);
+
+	struct PACKET_ZC_CASTLE_TELEPORT_RESPONSE p;
+	p.packetType = HEADER_ZC_CASTLE_TELEPORT_RESPONSE;
+	p.result = (int16)result;
+	clif_send(&p, sizeof(p), &sd->bl, SELF);
+#endif
+}
+
+static void clif_parse_guild_castle_info_request(int fd, struct map_session_data* sd)
+{
+#if PACKETVER_MAIN_NUM >= 20190522 || PACKETVER_RE_NUM >= 20190522 || PACKETVER_ZERO_NUM >= 20190515
+	const struct PACKET_CZ_CASTLE_INFO_REQUEST* p = (struct PACKET_CZ_CASTLE_INFO_REQUEST*)RFIFOP(fd, 0);
+	struct guild* g = sd->guild;
+
+	if (g == nullptr)
+		return;
+
+	std::shared_ptr<guild_castle> gc = castle_db.find(p->castle_id);
+	if (gc == nullptr)
+		return;
+	if (gc->guild_id != g->guild_id)
+		return;
+	clif_guild_castleinfo(sd, gc->castle_id, gc->economy, gc->defense);
+#endif
+}
+
+void clif_parse_guild_castle_teleport_request(int fd, struct map_session_data* sd)
+{
+#if PACKETVER_MAIN_NUM >= 20190522 || PACKETVER_RE_NUM >= 20190522 || PACKETVER_ZERO_NUM >= 20190515
+	const struct PACKET_CZ_CASTLE_TELEPORT_REQUEST* p = (struct PACKET_CZ_CASTLE_TELEPORT_REQUEST*)RFIFOP(fd, 0);
+	struct guild* g = sd->guild;
+
+	if (g == nullptr)
+		return;
+
+	std::shared_ptr<guild_castle> gc = castle_db.find(p->castle_id);
+
+	if (gc == nullptr)
+		return;
+	if (gc->enable_client_warp == false)
+		return;
+	if (gc->guild_id != g->guild_id)
+		return;
+
+	if (map_getmapflag(sd->bl.m, MF_GVG_CASTLE))
+		return;
```

Comment:
Maybe we should check if the player is not on a map with `MF_NOTELEPORT` or `MF_NOWARP` (or others?) mapflags. Otherwise, if I'm not wrong, a player in jail may use this feature?

Reply:
Thanks @Atemo, I added additional checks.